### PR TITLE
Jenkins-Mesos additions

### PR DESCRIPTION
**Pull request by eadasiak**
So this PR adds a few (hopefully non-destructive) files to the mix.  Here's how all this works:

- The .json files are Marathon app definitions.  I created them based on the various Makefile targets, along with a generic one I've been using for testing.  Right now the ports are left to be dynamically assigned, but we should pick fixed ports w/in the resource range of the slaves so it's consistent.  For parity with the Makefile, there are some environment variables in the .json that need to be statically set that are currently being dynamically passed from the environment in the Makefile.
- The Docker image in the Marathon definitions point to my public repo.  That should probably change once they've been vetted.  The jenkins-mesos image is a Jenkins instance with an nginx proxy to route requests to the /service/jenkins URI.  This image is the master of the Jenkins-mesos cloud.  It will, in turn, launch a Jenkins framework w/in the Mesos cluster.  Chamberlain was added to this image for completeness, but since it's set for 0 executors, it could probably be removed.
- At container launch, there's a bootstrap script that rewrites some of the Jenkins host/port information such that the slave executors know how to communicate back to the Jenkins master.
- The framework will launch tasks on the dind-slave image defined in the <clouds> config.  It launches a docker-in-docker container, then executes a wrapper.sh around the command being sent from the framework.  I've included the chamberlain installation on this image.  This also should be ultimately moved from my public repo.
- The plugins directory contains the mesos plugin, and the entire directory gets pulled into the Docker image.
- The jobs directory is also pulled straight into the Docker image, and is unmodified otherwise.
- The passing of the ssh keys will need to be figured out still (alas, one area left where persistence would be nice).

Somehow, I think I may have missed the mark with this, but hopefully we'll be able to get it working to fit the CI workflows needed.  Thanks!

### CODE REVIEW

File: dind-agent/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+# Populate JDK env
+ENV JAVA_HOME=/usr/lib/jvm/default-jvm
+ENV PATH=${PATH}:${JAVA_HOME}/bin
+
+# Add any SSH known hosts to the environment variable $SSH_KNOWN_HOSTS
+ENV SSH_KNOWN_HOSTS github.com
+RUN ssh-keyscan $SSH_KNOWN_HOSTS | tee /etc/ssh/ssh_known_hosts
+
+RUN git clone https://github.com/behance/chamberlain /opt/chamberlain \
```
Comment (chr0n1x) [inline dind-agent/Dockerfile]: why is chamberlain being cloned into the worker node/container?

File: jenkins.model.JenkinsLocationConfiguration.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (chr0n1x) [inline jenkins.model.JenkinsLocationConfiguration.xml]: what's this?

File: nodeMonitors.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (chr0n1x) [inline nodeMonitors.xml]: is this file actually needed?

File: config.xml.mesos [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      <master>zk://leader.mesos:2181/mesos</master>
+      <description></description>
+      <frameworkName>jenkins</frameworkName>
+      <slavesUser>root</slavesUser>
+      <principal>jenkins</principal>
+      <secret></secret>
+      <checkpoint>true</checkpoint>
+      <onDemandRegistration>false</onDemandRegistration>
+      <jenkinsURL></jenkinsURL>
+      <slaveInfos>
```
Comment (chr0n1x) [inline config.xml.mesos]: What types of instances did you run this on?
Also, if this config does not have anything sensitive, I would just add it to the actual `config.xml`

File: Dockerfile.mesos [hunk 5]
```diff
@@ -0,0 +1,24 @@
+FROM mesosphere/jenkins:0.2.0
+
+ENV JENKINS_WAR_URL https://updates.jenkins-ci.org/download/war/1.642.1/jenkins.war
+ENV JENKINS_STAGING /var/jenkins_staging
+ENV JENKINS_HOME /var/jenkins_home
+ENV JENKINS_FOLDER /usr/share/jenkins/
+ENV JAVA_HOME "/usr/lib/jvm/java-7-openjdk-amd64"
```
Comment (chr0n1x) [inline Dockerfile.mesos]: if you're already using `FROM mesosphere/jenkins:0.2.0`, why are you also setting these envs? They're already [here](https://github.com/mesosphere/jenkins-mesos/blob/master/Dockerfile#L13)

File: Dockerfile.mesos [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    apt-get install -y lsb-release && \
+    echo "deb http://repos.mesosphere.io/$(lsb_release -is | tr '[:upper:]' '[:lower:]') $(lsb_release -cs)  main" | tee /etc/apt/sources.list.d/mesosphere.list && \
+    apt-get update && \
+    apt-get install -y mesos build-essential python-pip && \
+    git clone https://github.com/behance/chamberlain /opt/chamberlain && \
+    cd /opt/chamberlain && \
+    make && \
+    cd -
+
+COPY config.xml.mesos "${JENKINS_HOME}/config.xml"
```
Comment (chr0n1x) [inline Dockerfile.mesos]: I'm actually ok with moving all of this into `Dockerfile`. We can then just say that we support mesos out-of-the box, but if they don't have a cluster then this just wont do anything.

File: jenkins.model.JenkinsLocationConfiguration.xml (hunk 2 above)
Comment (eadasiak) [inline jenkins.model.JenkinsLocationConfiguration.xml]: It's one of the files that gets modified via the bootstrap script.  The jenkinsUrl parameter is the important bit.  Here's an example of the file once the container's been launched:

<jenkins.model.JenkinsLocationConfiguration>
  <adminAddress>address not configured yet &lt;nobody@nowhere&gt;</adminAddress>
  <jenkinsUrl>http://192.168.99.100:11264</jenkinsUrl>

File: nodeMonitors.xml (hunk 3 above)
Comment (eadasiak) [inline nodeMonitors.xml]: Since the nodes are ephemeral, my sense is that we'd want to disable all the preventative monitoring, but maybe it's not actually required.  I'll check and see what the behavior is without it.

File: config.xml.mesos (hunk 4 above)
Comment (eadasiak) [inline config.xml.mesos]: I was running it locally via a docker-machine with a 2-CPU, 4GB RAM configuration (thus the verrry tiny CPU increments in the Marathon app definitions).  Let me check and see what the behavior is if the plugin is loaded and the Mesos libraries aren't installed.

File: Dockerfile.mesos (hunk 5 above)
Comment (eadasiak) [inline Dockerfile.mesos]: But adding them again makes them, ummm, better?

Good catch, I'll get rid of them.

File: dind-agent/Dockerfile (hunk 1 above)
Comment (eadasiak) [inline dind-agent/Dockerfile]: 99% chance of me being stupid, but since the master is set to not process jobs by default (executors is set to 0), I figured it would have to be run *somewhere*.  If chamberlain is already being run in a container, then I'll be more than happy to pull this out.  By adding it in, I managed to make an Alpine Linux-based image 500MB

File: nodeMonitors.xml (hunk 3 above)
Comment (eadasiak) [inline nodeMonitors.xml]: Ya, looks like this needs to stay in there.  Otherwise, monitoring for the nodes gets unhappy.

File: dind-agent/Dockerfile (hunk 1 above)
Comment (chr0n1x) [inline dind-agent/Dockerfile]: The master should only have 1 executor which is used to run maintenance jobs, such as self-update. All job templates will specify nodes other than `master` though (whatever you decide the mesos workers to be).

File: nodeMonitors.xml (hunk 3 above)
Comment (chr0n1x) [inline nodeMonitors.xml]: huh, I would've thought that it would just be regenerated :(

File: config.xml.mesos (hunk 4 above)
Comment (chr0n1x) [inline config.xml.mesos]: the mesos libs can still be installed via the dockerfile, no?

File: dind-agent/Dockerfile (hunk 1 above)
Comment (eadasiak) [inline dind-agent/Dockerfile]: Good point!  I've modified the config so that there's 1 executor on the master, and added a slave label of 'mesos' so that slave jobs can use that assignedNode value.

Comment (chr0n1x) [inline dind-agent/Dockerfile]: yup!

File: mesos/scripts/bootstrap.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python2
+"""
```
Comment (chr0n1x) [inline mesos/scripts/bootstrap.py]: woooooooooooooooooooooooooooooooooooooooooooooooooooooooow

**Comment by chr0n1x** [comment]
Ya know what. If it works, let's just merge and move forward trying to get this into IT :P

**Comment by chr0n1x** [comment]
:+1:
